backend/scripts/mirror_s3_data.sh

```diff
@@ -11,30 +11,119 @@ set -e
-$S3_SYNC_CMD s3://dataset-assets-public-${SRC_ENV}/ s3://dataset-assets-public-${DEST_ENV}/
-set +x
+AWS_OPTIONS="--copy-props metadata-directive --no-progress"
+S3_COPY_CMD="aws s3 cp $AWS_OPTIONS"
+S3_SYNC_CMD="aws s3 sync $AWS_OPTIONS"
+
+latest_snapshot=$(aws s3 cp s3://cellxgene-wmg-${SRC_ENV}/latest_snapshot_run -)
+if [[ $DEST_ENV == 'rdev' ]]; then
+  s3_destination="env-rdev-wmg/${STACK}"
+else
+  s3_destination="cellxgene-wmg-${DEST_ENV}"
+fi
+echo -n $latest_snapshot | aws s3 cp - s3://${s3_destination}/latest_snapshot_run
+$S3_SYNC_CMD s3://cellxgene-wmg-${SRC_ENV}/${latest_snapshot} s3://${s3_destination}/${latest_snapshot}
+
+
+echo Mirroring S3 Dataset data from $SRC_ENV to $DEST_ENV...
+
+if [[ $DEST_ENV == 'rdev' ]]; then
+  # The code in this block implements Collection subsetting for rdev stacks. For every Dataset (current and prior
+  # versions) in every Collection whose uuid is passed into the commma-separated $COLLECTIONS arg, a sample valid
+  # Dataset is copied to all relevant s3 locations:
+  # - s3://env-rdev-artifacts/${STACK}/
+  # - s3://env-rdev-cellxgene/${STACK}/
+  # - s3://env-rdev-datasets/${STACK}/
+
+  # Clean up semaphores state directory ahead of using sem in case it was not cleared properly before
+  [[ -d "~/.parallel/semaphores" ]] && rm -r ~/.parallel/semaphores
+  sample_data_name="schema_3_0_0"
+  echo "Copying sample data for S3 Dataset assets for $(tr ',' '\n' <<< $COLLECTIONS | wc -l) Collections..."
+
+  # Copy artifacts
+  export AWS_PROFILE=$src_aws_profile  # For SRC_ENV
+  DB_PW=$(aws secretsmanager get-secret-value --secret-id corpora/backend/${SRC_ENV}/database --region us-west-2 | jq -r '.SecretString | match(":([^:]*)@").captures[0].string')
+  # Select distinct local "directory" paths for assets, e.g. 's3://bucket/uuid/local.h5ad' becomes 's3://bucket/uuid/'
+  uri_query="select distinct regexp_replace(uri, '(^.*\\/).*', '\\1') from \"DatasetArtifact\" where id in (select unnest(artifacts) from \"DatasetVersion\" where id in (select unnest(datasets) from \"CollectionVersion\" where collection_id in ('$(sed "s/,/','/g" <<< $COLLECTIONS)')))"
+  DEPLOYMENT_STAGE=$SRC_ENV make db/tunnel/up
+  uris=($(PGOPTIONS='-csearch_path=persistence_schema' PGPASSWORD=${DB_PW} psql --dbname corpora_${SRC_ENV} --username corpora_${SRC_ENV} --host 0.0.0.0 --csv --tuples-only -c "$uri_query"))
+  export AWS_PROFILE=single-cell-dev  # For DEST_ENV
+
+  for uri in "${uris[@]}"; do
+    bucket=$(sed -E 's/s3:\/\/([^\/]+).*/\1/' <<< $uri)
+    if [[ -n `grep 'hosted-cellxgene' <<< $bucket` ]]; then
+      rdev_bucket_suffix="cellxgene"
+    elif [[ -n `grep 'corpora-data' <<< $bucket` ]]; then
+      rdev_bucket_suffix="artifacts"
+    fi
+    key_dir=$(sed -E 's/s3:\/\/([^\/]+)\/(.*)/\2/' <<< $uri)
+    if [[ -n "$DATA" ]]; then
+      sem -j+0 $S3_SYNC_CMD s3://${bucket}/${key_dir} s3://env-rdev-${rdev_bucket_suffix}/${STACK}/${key_dir}
```

Comment:
`sem` gives us concurrency, managing a queue of jobs for a fixed (default of 12 for all of us on standard macbooks) number of processes.

Reply:
ah, `GNU sem is an alias for GNU parallel --semaphore.`

Reply:
oh cool. I knew it use parallel under the hood but didn't realize you could invoke the paradigm from parallel itself